tensornetwork/tn_keras/entangler.py

```diff
@@ -0,0 +1,196 @@
+import tensorflow as tf
+from tensorflow.keras.layers import Layer  # type: ignore
+from tensorflow.keras import activations
+from tensorflow.keras import initializers
+from typing import List, Optional, Text, Tuple
+import tensornetwork as tn
+from tensornetwork.network_components import Node
+import numpy as np
+import math
+
+
+@tf.keras.utils.register_keras_serializable()  # type: ignore
+class DenseEntangler(Layer):
+  """Entangler TN layer. Allows for very large hidden layers.
+
+  Example:
+
+  ```python
+  # as first layer in a sequential model:
+  model = Sequential()
+  model.add(
+    DenseEntangler(16**3,
+                    num_legs=3,
+                    num_levels=3,
+                    use_bias=True,
+                    activation='relu',
+                    input_shape=(16**3,)))
+  # now the model will take as input arrays of shape (*, 4096)
+  # and output arrays of shape (*, 4096).
+  # After the first layer, you don't need to specify
+  # the size of the input anymore:
+  model.add(DenseEntangler(16**3, num_legs=3, num_levels=3, use_bias=True))
+  ```
+
+  Args:
+    output_dim: Positive integer, dimensionality of the output space.
+      Note: output_dim must be equal to the dimensionality of the input.
+    num_legs: Positive integer, number of legs the state node has.
+    num_levels: Positive integer, number of levels we want the entangler to
+      have. This is the only parameter that does not change input/output shape.
+      It can be increased to increase the power of the layer, but inference
+      time will also scale approximately linearly.
+    activation: Activation function to use.
+      If you don't specify anything, no activation is applied
+      (ie. "linear" activation: `a(x) = x`).
+    use_bias: Boolean, whether the layer uses a bias vector.
+    kernel_initializer: Initializer for the two weight matrices.
+    bias_initializer: Initializer for the bias vector.
+
+  Input shape:
+    2D tensor with shape: `(batch_size, input_shape[-1])`.
+
+  Output shape:
+    2D tensor with shape: `(batch_size, output_dim)`.
+  """
+
+  def __init__(self,
+               output_dim: int,
+               num_legs: int,
+               num_levels: int,
+               use_bias: Optional[bool] = True,
+               activation: Optional[Text] = None,
+               kernel_initializer: Optional[Text] = 'glorot_uniform',
+               bias_initializer: Optional[Text] = 'zeros',
+               **kwargs) -> None:
+
+    if 'input_shape' not in kwargs and 'input_dim' in kwargs:
+      kwargs['input_shape'] = (kwargs.pop('input_dim'),)
+
+    assert num_legs > 2, 'Need at least 2 legs to create Entangler'
```

Comment:
Can you add a printout of what the received value was here?

Reply:
Updated!